src/lib/mongoBackend/MongoCommonUpdate.cpp

```diff
@@ -2645,12 +2936,8 @@ void processContextElement
   //
   // FIXME P6: Once we allow for ServicePath to be modified, this loop must be looked at.
   //
-  int docs = 0;
-
-  // Used to accumulate error response information, checked at the end
-  bool         attributeAlreadyExistsError = false;
-  std::string  attributeAlreadyExistsList  = "[ ";
 
+  std::vector<BSONObj> results;
   while (moreSafe(cursor))
```

Comment:
Indeed, not sure if that is fully right... As far as I see, there is a trade-off.

On the one side, the two-phase approach will release before the connection, so it would minimize the time waiting at it. On the other side, two-phase approach implies that some work is done twice versus the one-phase approach, which do all the work in the same pass.

Let's consider the following case: a queryContext that results in 1,000 elements (the maximum according to hard pagination limit). The one phase approach works in this way:

1. Take sem
2. Process 1,000 results from cursor
3. Release sem

The two-phase appraoch works this way:

1. Taken sem
2. Create a 1,000 BSONObj vector
3. Release sem
4. Process 1,000 results from vector

What would be the time from 1 to 3 in both cases? How compares the time in one-phase step 2 versus the sum of the times in two-phases 2 and 4? What about the impact of the 1,000 BSONObjt vector structure (x N, given that N threads at a time could be in step 4) compared with the cursor, which doesn't all the results at a time but gets them from MongoDB batch by batch throught the wiring protocol?

However, leaving performance appart. There is an good reason to use the two-phase approach: it would allow a thread-safe interface for all colletionQuerry() invokations. I mean, we could use the following alternative for collectionQuery():

```
extern bool collectionQuery
(
  const std::string&              col,
  const BSONObj&                  q,
  std::vector<BSONObj>*     results,
  std::string*                    err
);
```

and the connectionQuery() caller wouldn't need to deal with getMongoConnection()/releaseMongoConnection() or TIME_STAT_MONGO_READ_WAIT_START/STOP. In fact, the caller of that hypotecial collectionQuery() function wouldn't ever see the cursor. That's would lead to cleaner and safer code.

However, given that 0.26.0 closing is very near and that a decission on this needs to be analyzed carefully (there are many aspect to consider) I think is a good idea to spin off to a future issue (basically a copy-paste of this comment :)